XDMA/linux-kernel/xdma/libxdma.c

```diff
@@ -1375,61 +1382,133 @@ static irqreturn_t xdma_isr(int irq, void *dev_id)
+
+		/* Start the engine */
+		channel_interrupts_enable(engine->xdev, engine->irq_bitmask);
+		iowrite32(DMA_ENGINE_START, &engine->regs->control);
 	}
 
 	mask = ch_irq & xdev->mask_irq_h2c;
 	if (mask) {
-		int channel = 0;
-		int max = xdev->h2c_channel_max;
-
-		/* iterate over H2C (PCIe read) */
-		for (channel = 0; channel < max && mask; channel++) {
-			struct xdma_engine *engine = &xdev->engine_h2c[channel];
-
-			/* engine present and its interrupt fired? */
-			if ((engine->irq_bitmask & mask) &&
-			    (engine->magic == MAGIC_ENGINE)) {
-				mask &= ~engine->irq_bitmask;
-				dbg_tfr("schedule_work, %s.\n", engine->name);
-				schedule_work(&engine->work);
-			}
-		}
-	}
+		struct xdma_private *priv = netdev_priv(ndev);
+		struct xdma_engine *engine = &xdev->engine_h2c[0];
+		struct xdma_desc *desc;
+		struct sk_buff *skb;
+		unsigned long flag;
+		int index;
+		int desc1_status = 0;
+		int desc2_status = 0;
+		dma_addr_t bus_addr;
+		u32 w;
+
+		/* Read and clear the desc register */
+		engine_status_read(engine, 1, 0);
+
+		/* Check last used descriptor */
+		spin_lock_irqsave(&priv->tx_lock, flag);
+		index = priv->last;
+		iowrite32(DMA_ENGINE_STOP, &engine->regs->control);
+		spin_unlock_irqrestore(&priv->tx_lock, flag);
 
-	mask = ch_irq & xdev->mask_irq_c2h;
-	if (mask) {
-		int channel = 0;
-		int max = xdev->c2h_channel_max;
-
-		/* iterate over C2H (PCIe write) */
-		for (channel = 0; channel < max && mask; channel++) {
-			struct xdma_engine *engine = &xdev->engine_c2h[channel];
-
-			/* engine present and its interrupt fired? */
-			if ((engine->irq_bitmask & mask) &&
-			    (engine->magic == MAGIC_ENGINE)) {
-				mask &= ~engine->irq_bitmask;
-				dbg_tfr("schedule_work, %s.\n", engine->name);
-				schedule_work(&engine->work);
+		channel_interrupts_enable(engine->xdev, engine->irq_bitmask);
+
+		if (index == 0) {
+			/* Initialize last descriptor */
+			spin_lock_irqsave(&priv->desc_lock[0], flag);
+			desc = priv->desc[0];
+			desc->src_addr_hi = 0;
+			desc->src_addr_lo = 0;
+			skb = priv->skb[0];
+			spin_unlock_irqrestore(&priv->desc_lock[0], flag);
+
+			/* Check if the other descriptor is ready */
+			spin_lock_irqsave(&priv->desc_lock[1], flag);
+			desc = priv->desc[1];
+			if (desc->src_addr_hi != 0 && desc->src_addr_lo != 0) {
+				desc2_status = DESC_READY;
 			}
+			spin_unlock_irqrestore(&priv->desc_lock[1], flag);
+		} else if (index == 1) {
+			/* Initialize last descriptor */
+			spin_lock_irqsave(&priv->desc_lock[1], flag);
+			desc = priv->desc[1];
+			desc->src_addr_hi = 0;
+			desc->src_addr_lo = 0;
+			skb = priv->skb[1];
+			spin_unlock_irqrestore(&priv->desc_lock[1], flag);
+			
+			/* Check if the other descriptor is ready */
+			desc = priv->desc[0];
+			spin_lock_irqsave(&priv->desc_lock[0], flag);
+			if (desc->src_addr_hi != 0 && desc->src_addr_lo != 0) {
+				desc1_status = DESC_READY;
+			}
+			spin_unlock_irqrestore(&priv->desc_lock[0], flag);
+		}
```

Comment:
This block is almost same as the block above it.
How about create a simple function for it and call it with `index` as a parameter